### PR TITLE
Add MultiMove support

### LINKED ISSUE #2: How to integrate an external axis?
We are interested in integrating an external axis (aka MultiMove). I did some reading at the old [ROS1 driver](https://github.com/ros-industrial/abb_robot_driver) and learned a few things:

- The `abb_rws_state_publisher` package publishes the JointState and SystemState, including external axis info.
- `abb_egm_hardware_interface` provides actual control of the axes. It has one EGM channel per "mechanical unit group."

Question:  there's a note saying `abb_egm_hardware_interface` relies on `abb_rws_service_provider`. Is that a hard requirement or something we could avoid?

Related:  https://github.com/ros-industrial/abb_libegm/issues/57

### PR DESCRIPTION
**Pull request by stephanie-eng**
Switches from using abblibegm's EGMControllerInterface to abb_egm_rws_manager's egmManager to get support for multiple mechanical unit groups.

Fixes #2

### CODE REVIEW

File: src/abb_system_position_only.cpp [hunk 1]
```diff
@@ -13,242 +13,200 @@
...
+  
+  // Create channel configuration for each mechanical unit group
+  std::vector<abb::robot::EGMManager::ChannelConfiguration> channel_configurations;
+  for (const auto & group : robot_controller_description_.mechanical_units_groups()) {
+    // Each channel needs a new port number, start with port from hardware parameters and increment
+    // by 1 for each channel
+    auto channel_configuration =
+      abb::robot::EGMManager::ChannelConfiguration{static_cast<uint16_t>(robotstudio_port), group};
+    channel_configurations.emplace_back(channel_configuration);
+    robotstudio_port += 1;
```
Comment (stephanie-eng) [inline src/abb_system_position_only.cpp]: I couldn't find a way to get the remote port number for each mechanical unit group. For now, we can perhaps assume sequentially-increasing port numbers, but I hope there is a more correct way to do this.

File: src/abb_system_position_only.cpp [hunk 2]
```diff
@@ -13,242 +13,200 @@
...
     return CallbackReturn::ERROR;
   }
 
-  hw_states_.resize(info_.joints.size(), std::numeric_limits<double>::quiet_NaN());
-  hw_commands_.resize(info_.joints.size(), std::numeric_limits<double>::quiet_NaN());
-  robotstudio_port_ = stod(info_.hardware_parameters["robotstudio_port"]);
+  // TODO(seng):
+  // Rename robotstudio_port to egm_port
+  // Create new hardware parameter for:
+  //  - rws_port (8000)
```
Comment (stephanie-eng) [inline src/abb_system_position_only.cpp]: RWS by default runs on port 80, which is a reserved port on Linux. This leaves us with 3 options that I can think of (I intend to add a section in the docs on how to do configure these any of these options regardless):

1. Require users to change the RWS port in RobotStudio to a different port
2. Leave the default RWS port as 80, and use port forwarding to listen to port 80 on the RobotStudio machine and forward it to a different port on the remote machine
3. Require the user to use sudo when setting up the port forward to use port 80 on both the local and remote machines

Depending on which we think is best, we should set the hardware parameter for the RWS port to either 80 or 8000.

Comment (gavanderhoorn) [reply src/abb_system_position_only.cpp]: `tcp://:80` is used by the RWS server.

Your driver is the client.

Which port you use locally does not matter to the server.

File: src/abb_system_position_only.cpp (hunk 1 above)
Comment (gavanderhoorn) [reply src/abb_system_position_only.cpp]: This depends on the EGM configuration on the controller.

This is a bit of information your user would have to provide.

File: src/abb_system_position_only.cpp (hunk 2 above)
Comment (stephanie-eng) [reply src/abb_system_position_only.cpp]: I agree with your statements, but I'd like to choose a default local port number and I'm wondering what local port would be simplest to use. 

My question isn't really a big deal, as ideally a user will change the IP and port to match their setup, but it would be nice to have sane default values. I can just leave it the local port as 80, unless you have any objections. This would work for people that configure their virtual controller whitelist for their local machine, but may cause issue as it's a reserved port on Linux.

Comment (gavanderhoorn) [reply src/abb_system_position_only.cpp]: > I can just leave it the local port as 80

I'm not sure I understand what you're writing.

The IRC5 controller (or the virtual one in RobotStudio) is the server. The RWS HTTP server running on the IRC uses `tcp://:80` to accept requests from clients.

Your driver is one of those clients.

The client will use a random, local, TCP port, assigned by the OS, to connect to the server.

There should be no need to "choose a default local port number", your OS will choose one for you.

Setting the `rws_ip` to `127.0.0.1` doesn't really make sense to me. Was port-forwarding used for some reason? That's not necessary, if the IP whitelist has been correctly configured (https://github.com/ros-industrial/abb_librws/issues/81). I would recommend to *not* make that work-around the default, and just assume the driver is connecting to real hw.

Comment (stephanie-eng) [reply src/abb_system_position_only.cpp]: I wasn't able to get the whitelisting to work for some reason, so I've been using port-forwarding as a workaround. I understand your point that port-forwarding shouldn't be the default - I've changed the code such that `rws_ip` is specified from a launch parameter, as it should never be localhost if whitelisting is working correctly.

Appreciate your help!

File: abb.repos [hunk 3]
```diff
@@ -1,5 +1,13 @@
 repositories:
   abb_libegm:
     type: git
-    url: https://github.com/ros-industrial/abb_libegm.git
+    url: git@github.com:ros-industrial/abb_libegm.git
```
Comment (AndyZe) [inline abb.repos]: I don't think this is correct for the average person. I'm guessing it works for you because you use ssh keys? See https://github.com/ros-planning/moveit2/blob/main/moveit2.repos

File: docs/README.md [hunk 4]
```diff
@@ -87,14 +92,105 @@ The simulation will then try to connect with the ROS2 driver every few seconds.
...
+
+There should be a line that says `Listen -1` line - change the `-1` to your desired port number and save the file.
+
+To verify the port used by RWS, open Task Manager and click the `Details` tab. Look for `RobVC.exe` and note the PID. Open command prompt, and run the following command, substituting `<PID_NUMBER>` with the PID from earlier.
+
+`netstat -aon | findStr <PID NUMBER>`
+
+You should see something like the following - in this case, RWS is listening on port 80:
+
+![netstat](images/netstat.PNG)
```
Comment (AndyZe) [inline docs/README.md:135]: Please move this image into the `docs` folder

File: docs/README.md [hunk 5]
```diff
@@ -52,7 +52,12 @@ These steps cover how to set up a new robot in RobotStudio to work with the ABB
 
 2. In the window that pops up for controller customization, add EGM to the controller. This can be done by selecting `Engineering Tools` on the sidebar, and selecting the following options:
 - `689-1 Externally Guided Motion (EGM)`
-- (not sure if needed) `623-1 Multitasking`
+
+For MultiMove support, also select the following
```
Comment (AndyZe) [inline docs/README.md]:
```suggestion
For MultiMove support, e.g. for external axes, also select the following:
```

File: docs/README.md [hunk 6]
```diff
@@ -52,7 +52,12 @@ These steps cover how to set up a new robot in RobotStudio to work with the ABB
...
+- `623-1 Multitasking` also under `Engineering Tools`
+- `604-1 MultiMove Independent` 
+
+Note that EGM Position Stream does not support MultiMove Coordinated.
```
Comment (AndyZe) [inline docs/README.md]: Can you add an explanation for this in plain English? There are too many marketing terms for me to follow. This probably isn't correct but here's an example:

```suggestion
Note that EGM Position Stream does not support MultiMove Coordinated. This means that external axis motion, for example, will not be synchronized with robot motion
```

File: docs/README.md [hunk 7]
```diff
@@ -87,14 +92,105 @@ The simulation will then try to connect with the ROS2 driver every few seconds.
 
 ## Connecting with ROS2
 
-To launch with RobotStudio, set `use_fake_hardware:=false`. As far as ROS is aware, RobotStudio is a real robot:
+The driver reqiures some network information to connect to EGM and RWS. This information is store in the ros2_control [robot description file](../robot_specific_config/abb_irb1200_support/urdf/irb1200.ros2_control.xacro). Change the following lines as required:
```
Comment (AndyZe) [inline docs/README.md]:
```suggestion
The driver reqiures some network information to connect to EGM and RWS. This information is stored in the ros2_control [robot description file](../robot_specific_config/abb_irb1200_support/urdf/irb1200.ros2_control.xacro). Change the following lines as required:
```

File: docs/README.md [hunk 8]
```diff
@@ -87,14 +92,105 @@ The simulation will then try to connect with the ROS2 driver every few seconds.
...
+<param name="rws_port">80</param>
+<param name="rws_ip">127.0.0.1</param>
+<param name="rob1egm_port">6511</param>
+```
+
+- `rws_port` should be the port for RWS
+- `rws_ip` should be the IP address of the RobotStudio computer (see [Adding the local computer to the virtual controller whitelist](#adding-the-local-computer-to-the-virtual-controller-whitelist))
+- `<MECHANICAL_UNIT_GROUP_NAME>egm_port` should match the `Remote port` setting from the RobotStudio controller configuration
+     - Change `<MECHANICAL_UNIT_GROUP_NAME>` to match the mechanical unit group name - in the sample Pack and Go file, this is `rob1`
+     - If not using MultiMove, change the paramter name to be `egm_port`
```
Comment (AndyZe) [inline docs/README.md]:
```suggestion
     - If not using MultiMove, change the parameter name to `egm_port`
```

File: docs/README.md [hunk 9]
```diff
@@ -87,14 +92,105 @@ The simulation will then try to connect with the ROS2 driver every few seconds.
...
-    ros2 launch abb_bringup abb_control.launch.py description_package:=abb_irb1200_support description_file:=irb1200_5_90.xacro launch_rviz:=false moveit_config_package:=abb_irb1200_5_90_moveit_config
+    ros2 launch abb_bringup abb_control.launch.py description_package:=abb_irb1200_support description_file:=irb1200_5_90.xacro launch_rviz:=false moveit_config_package:=abb_irb1200_5_90_moveit_config use_fake_hardware:=false robotstudio_ip:=<ROBOTSTUDIO_IP>
 
 After launching the controllers, launch MoveIt:
 
     ros2 launch abb_bringup abb_moveit.launch.py robot_xacro_file:=irb1200_5_90.xacro support_package:=abb_irb1200_support
 
+## Connecting to RWS
+
+RWS is a platform that allows interaction with the robot controller over HTTP, which is used by the driver to get information about the robot. Note that only RWS 1.0 is supported - RobotWare 7.0 and higher use RWS 2.0 and are not currently supported.
```
Comment (AndyZe) [inline docs/README.md:119]: > RobotWare 7.0 and higher use RWS 2.0 and are not currently supported.

This makes me sad. Oh well, maybe later

File: docs/README.md [hunk 10]
```diff
@@ -87,14 +92,105 @@ The simulation will then try to connect with the ROS2 driver every few seconds.
...
+
+![netstat](images/netstat.PNG)
+
+Note that if you change the RWS port, you may need to change the `rws_port` parameter in the [robot description file](../robot_specific_config/abb_irb1200_support/urdf/irb1200.ros2_control.xacro) to match the port you selected.
+
+See [this forum post](https://forums.robotstudio.com/discussion/12177/how-to-change-the-listening-port-of-the-virtual-controller-robotware-6-x-and-7-x) for more detail.
+
+### Adding the local computer to the virtual controller whitelist
+
+To access the virtual controller with RWS from a networked ROS2 computer, you will need to whitelist the ROS2 computer's IP on the RobotStudio computer. Without this whitelist, if you will get an RAPI Unidentified Error when trying to access the controller. You can verify this if you try to navigate to `<ROBOTSTUDIO_COMPUTER_IP>:80` from the ROS2 computer. The default credentials are:
```
Comment (AndyZe) [inline docs/README.md]:
```suggestion
To access the virtual controller with RWS from a networked ROS2 computer, you will need to whitelist the ROS2 computer's IP on the RobotStudio computer. Without this whitelist, you will get an RAPI Unidentified Error when trying to access the controller. You can verify this if you try to navigate to `<ROBOTSTUDIO_COMPUTER_IP>:80` from the ROS2 computer. The default credentials are:
```

File: docs/README.md [hunk 11]
```diff
@@ -87,14 +92,105 @@ The simulation will then try to connect with the ROS2 driver every few seconds.
...
+To launch with RobotStudio, set `use_fake_hardware:=false` and `robotstudio_ip:=<ROBOTSTUDIO_IP>`, substituting `<ROBOTSTUDIO_IP>` with the IP of the RobotStudio computer. As far as ROS is aware, RobotStudio is a real robot:
 
-    ros2 launch abb_bringup abb_control.launch.py description_package:=abb_irb1200_support description_file:=irb1200_5_90.xacro launch_rviz:=false moveit_config_package:=abb_irb1200_5_90_moveit_config
+    ros2 launch abb_bringup abb_control.launch.py description_package:=abb_irb1200_support description_file:=irb1200_5_90.xacro launch_rviz:=false moveit_config_package:=abb_irb1200_5_90_moveit_config use_fake_hardware:=false robotstudio_ip:=<ROBOTSTUDIO_IP>
 
 After launching the controllers, launch MoveIt:
 
     ros2 launch abb_bringup abb_moveit.launch.py robot_xacro_file:=irb1200_5_90.xacro support_package:=abb_irb1200_support
 
+## Connecting to RWS
```
Comment (AndyZe) [inline docs/README.md:117]: IS RWS 100% mandatory now, or only for external axes? I don't like how it makes the README more complicated.

File: abb_hardware_interface/src/utilities.cpp [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+namespace abb
+{
+namespace robot
+{
+namespace utilities
+{
+/***********************************************************************************************************************
+ * Utility function definitions
+ */
```
Comment (AndyZe) [inline abb_hardware_interface/src/utilities.cpp]:
```suggestion
```

We usually don't add comments like this. Maybe you should put a description at the top of the file? Although the filename makes it pretty obvious, so maybe that's not even needed

File: abb_hardware_interface/src/utilities.cpp [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+  "Failed to establish RWS connection to the robot controller"};
+
+/**
+ * \brief Time [s] to wait before trying to reconnect to a robot controller via RWS.
+ */
+constexpr uint8_t RWS_RECONNECTION_WAIT_TIME{1};
+auto LOGGER = rclcpp::get_logger("ABBSystemPositionOnlyHardware");
+}  // namespace
+
+namespace abb
```
Comment (AndyZe) [inline abb_hardware_interface/src/utilities.cpp]: The outer namespace should be the package name, `abb_ros2`

File: abb_hardware_interface/src/utilities.cpp [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+ * \brief Error message for failed connection attempts when trying to connect to a robot controller via RWS.
+ */
+constexpr char RWS_CONNECTION_ERROR_MESSAGE[]{
+  "Failed to establish RWS connection to the robot controller"};
+
+/**
+ * \brief Time [s] to wait before trying to reconnect to a robot controller via RWS.
+ */
+constexpr uint8_t RWS_RECONNECTION_WAIT_TIME{1};
+auto LOGGER = rclcpp::get_logger("ABBSystemPositionOnlyHardware");
```
Comment (AndyZe) [inline abb_hardware_interface/src/utilities.cpp]:
```suggestion
auto LOGGER = rclcpp::get_logger("ABBHardwareInterfaceUtilities");
```

File: abb_hardware_interface/src/utilities.cpp [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+ ***********************************************************************************************************************
+ */
+
+#include <abb_hardware_interface/utilities.hpp>
+
+#include <stdexcept>
+
+#include "rclcpp/rclcpp.hpp"
+
+namespace
```
Comment (AndyZe) [inline abb_hardware_interface/src/utilities.cpp:52]: I think this anonymous namespace should go inside the other namespaces.

File: abb_hardware_interface/src/abb_system_position_only.cpp [hunk 16]
```diff
@@ -12,243 +12,199 @@
...
-      hw_commands_[i], i);
-    output_pos_.mutable_robot()->mutable_joints()->mutable_position()->set_values(
-      i, (hw_commands_[i] * 180.0 / 3.14159));
-  }
-  egm_interface_->write(output_pos_);
+  RCLCPP_INFO(LOGGER, "write()");
 
+  for (auto & group : motion_data_.groups) {
+    for (auto & unit : group.units) {
+      for (auto & joint : unit.joints) {
```
Comment (AndyZe) [inline abb_hardware_interface/src/abb_system_position_only.cpp]: Is this too nitpicky? There are some other places that could use the same change.

```suggestion
  for (const auto & group : motion_data_.groups) {
    for (const auto & unit : group.units) {
      for (const auto & joint : unit.joints) {
```

File: abb_hardware_interface/src/abb_system_position_only.cpp [hunk 17]
```diff
@@ -12,243 +12,199 @@
...
-
-  for (size_t i = 0; i < hw_commands_.size(); ++i) {
-    RCLCPP_INFO(
-      rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "Got command %.5f for joint %ld!",
-      hw_commands_[i], i);
-    output_pos_.mutable_robot()->mutable_joints()->mutable_position()->set_values(
-      i, (hw_commands_[i] * 180.0 / 3.14159));
-  }
-  egm_interface_->write(output_pos_);
+  RCLCPP_INFO(LOGGER, "write()");
```
Comment (AndyZe) [inline abb_hardware_interface/src/abb_system_position_only.cpp]: Delete if done debugging

File: abb_hardware_interface/src/abb_system_position_only.cpp [hunk 18]
```diff
@@ -12,243 +12,199 @@
...
+  if (!egm_manager_->read(motion_data_)) {
+    RCLCPP_INFO(LOGGER, "Failed to read EGM feedback - is EGM session active?");
+  } else {
+    for (auto & group : motion_data_.groups) {
+      for (auto & unit : group.units) {
+        for (auto & joint : unit.joints) {
+          RCLCPP_INFO(LOGGER, "Got state %.5f for %s!", joint.state.position, joint.name.c_str());
+        }
+      }
+    }
```
Comment (AndyZe) [inline abb_hardware_interface/src/abb_system_position_only.cpp]: This looks like debug code that should be deleted

File: abb_hardware_interface/src/abb_system_position_only.cpp [hunk 19]
```diff
@@ -12,243 +12,199 @@
...
-  RCLCPP_INFO(rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "System Sucessfully started!");
-
-  return CallbackReturn::SUCCESS;
-}
 
-CallbackReturn ABBSystemPositionOnlyHardware::on_deactivate(
-  const rclcpp_lifecycle::State & /* previous_state */)
-{
-  RCLCPP_INFO(rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "on_deactivate()");
+  RCLCPP_INFO(LOGGER, "CONNECTED!");
```
Comment (AndyZe) [inline abb_hardware_interface/src/abb_system_position_only.cpp]: delete

File: abb_hardware_interface/src/abb_system_position_only.cpp [hunk 20]
```diff
@@ -12,243 +12,199 @@
...
 
-  io_service_.stop();
-  thread_group_.join_all();
+  egm_manager_->read(motion_data_);
 
-  RCLCPP_INFO(rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "System sucessfully stopped!");
+  RCLCPP_INFO(LOGGER, "System sucessfully started!");
```
Comment (AndyZe) [inline abb_hardware_interface/src/abb_system_position_only.cpp]: Tyler says, "don't log the happy path" but I think we can keep this. A little terminal output every once in awhile is a good thing. Let's make it more specific:

```suggestion
  RCLCPP_INFO(LOGGER, "ros2_control hardware interface was successfully started!");
```

File: abb_hardware_interface/src/abb_system_position_only.cpp [hunk 21]
```diff
@@ -12,243 +12,199 @@
...
   }
   return state_interfaces;
 }
 
-std::vector<hardware_interface::CommandInterface>
-ABBSystemPositionOnlyHardware::export_command_interfaces()
+std::vector<hardware_interface::CommandInterface> ABBSystemPositionOnlyHardware::export_command_interfaces()
 {
-  RCLCPP_INFO(rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "export_command_interfaces()");
+  RCLCPP_INFO(LOGGER, "export_command_interfaces()");
```
Comment (AndyZe) [inline abb_hardware_interface/src/abb_system_position_only.cpp]:
```suggestion
```

File: abb_hardware_interface/src/abb_system_position_only.cpp [hunk 22]
```diff
@@ -12,243 +12,199 @@
...
 }
 
-CallbackReturn ABBSystemPositionOnlyHardware::on_activate(
-  const rclcpp_lifecycle::State & /* previous_state */)
+CallbackReturn ABBSystemPositionOnlyHardware::on_activate(const rclcpp_lifecycle::State & /* previous_state */)
 {
-  RCLCPP_INFO(rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "on_activate()");
-
-  thread_group_.create_thread(boost::bind(&boost::asio::io_service::run, &io_service_));
+  RCLCPP_INFO(LOGGER, "on_activate()");
```
Comment (AndyZe) [inline abb_hardware_interface/src/abb_system_position_only.cpp]:
```suggestion
```

File: abb_hardware_interface/src/abb_system_position_only.cpp [hunk 23]
```diff
@@ -12,243 +12,199 @@
...
 
-  // done
-  return CallbackReturn::SUCCESS;
-}
-
-CallbackReturn ABBSystemPositionOnlyHardware::on_configure(
-  const rclcpp_lifecycle::State & /* previous_state */)
-{
-  RCLCPP_INFO(rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "on_configure()");
+  // Done
```
Comment (AndyZe) [inline abb_hardware_interface/src/abb_system_position_only.cpp]:
```suggestion
```

File: abb_hardware_interface/src/abb_system_position_only.cpp [hunk 24]
```diff
@@ -12,243 +12,199 @@
...
 namespace abb_hardware_interface
 {
 static constexpr size_t NUM_CONNECTION_TRIES = 100;
-static constexpr auto DEG_TO_RAD = 180 / M_PI;
+static const rclcpp::Logger LOGGER = rclcpp::get_logger("ABBSystemPositionOnlyHardware");
 
 CallbackReturn ABBSystemPositionOnlyHardware::on_init(const hardware_interface::HardwareInfo & info)
 {
-  RCLCPP_INFO(rclcpp::get_logger("ABBSystemPositionOnlyHardware"), "on_init()");
+  RCLCPP_INFO(LOGGER, "on_init()");
```
Comment (AndyZe) [inline abb_hardware_interface/src/abb_system_position_only.cpp]:
```suggestion
```

File: abb_hardware_interface/package.xml [hunk 25]
```diff
@@ -17,7 +17,7 @@
   <depend>hardware_interface</depend>
   <depend>pluginlib</depend>
   <depend>rclcpp</depend>
-  <depend>abb_libegm</depend>
+  <depend>abb_egm_rws_managers</depend>
```
Comment (AndyZe) [inline abb_hardware_interface/package.xml]: We still need abb_libegm, right?

(Please alphabetize, too)

File: docs/README.md (hunk 4 above)
Comment (stephanie-eng) [reply docs/README.md:135]: Not quite sure what you're asking here - is there a reason we want this image separate from the other images?

Comment (AndyZe) [reply docs/README.md:135]: Ah, nevermind. The way github presented the folder layout was misleading.

File: abb_hardware_interface/package.xml (hunk 25 above)
Comment (stephanie-eng) [reply abb_hardware_interface/package.xml]: We no longer use abb_libegm directly, we get that functionality via abb_egm_rws_managers (which depends on abb_libegm)

File: abb_hardware_interface/src/utilities.cpp (hunk 12 above)
Comment (stephanie-eng) [reply abb_hardware_interface/src/utilities.cpp]: This is a copy of https://github.com/ros-industrial/abb_robot_driver/blob/master/abb_robot_cpp_utilities/src/initialization.cpp, so we should probably link that at the top

File: docs/README.md (hunk 11 above)
Comment (stephanie-eng) [reply docs/README.md:117]: This PR currently makes RWS mandatory for all cases - RWS is used to get the robot setup from RobotStudio. We might be able to hardcode the correct configuration for a single robot with no external axis if necessary, but if the issue is with README length, connecting to RWS with the default settings shouldn't be very complicated.

A lot of the README is extra information on how to change configurations that are not strictly necessary and can be removed... I'm of the opinion that it's nice to have all that relevant information grouped together somewhere, but it doesn't necessarily have to be here.

File: abb_hardware_interface/src/utilities.cpp (hunk 13 above)
Comment (stephanie-eng) [reply abb_hardware_interface/src/utilities.cpp]: I'm fine to go either way, but this namespace convention is used by the ROS 1 abb driver (where this file was taken from), `abb_libegm` and `abb_egm_rws_managers`. We don't need to keep it that way if you feel that this should be changed, though.

File: docs/README.md (hunk 9 above)
Comment (stephanie-eng) [reply docs/README.md:119]: The RWS restriction is from abb_librws.
There is some talk about this here: https://github.com/ros-industrial/abb_librws/issues/147
It sounds like there is a fork with RWS 2 support - for now, I think we should just stick to what we have for simplicity, unless we have a need to use RWS 2 now. With any luck, the changes will be upstreamed.

File: docs/README.md [hunk 26]
```diff
@@ -52,7 +52,12 @@ These steps cover how to set up a new robot in RobotStudio to work with the ABB
 
 2. In the window that pops up for controller customization, add EGM to the controller. This can be done by selecting `Engineering Tools` on the sidebar, and selecting the following options:
 - `689-1 Externally Guided Motion (EGM)`
-- (not sure if needed) `623-1 Multitasking`
+
+For MultiMove support, e.g. for external axes, also select the following:
+- `623-1 Multitasking` also under `Engineering Tools`
+- `604-1 MultiMove Independent` 
+
+Note that EGM Position Stream does not support MultiMove Coordinated - this means that coordinated, synchronized robot movement is not supported (i.e. for tasks that involved multiple robots that are working on the same moving work object).
```
Comment (AndyZe) [inline docs/README.md]:
```suggestion
Note that EGM Position Stream does not support MultiMove Coordinated - this means that coordinated, synchronized robot movement is not supported (i.e. for tasks that involve multiple robots working on the same moving work object).
```

File: docs/README.md [hunk 27]
```diff
@@ -87,18 +92,109 @@ The simulation will then try to connect with the ROS2 driver every few seconds.
...
+```
+<VCConfiguration>
+<RemoteVCConfiguration PublicationEnabled="true" />
+<hosts>
+<host ip="ROS2_COMPUTER_IP"/>
+</hosts>
+</VCConfiguration>
+```
+
+Change ROS2_COMPUTER_IP to the IP of the ROS2 computer. Up to 10 IP addresses can be whitelisted by adding additional `<host ip="..."/>` lines.
```
Comment (AndyZe) [inline docs/README.md]:
```suggestion
Change ROS2_COMPUTER_IP to the IP of the ROS2 computer. Up to 10 IP addresses can be whitelisted by adding additional `<host ip="..."/>` lines. Restart RobotStudio after the update.
```

File: docs/README.md [hunk 28]
```diff
@@ -87,18 +92,109 @@ The simulation will then try to connect with the ROS2 driver every few seconds.
...
+
+See [this forum post](https://forums.robotstudio.com/discussion/12082/using-robotwebservices-to-access-a-remote-virtual-controller) for more details.
+
+### Configuring SSH port forwarding
+
+If you'd prefer to avoid the [whitelisting solution](#adding-the-local-computer-to-the-virtual-controller-whitelist), you can also use ssh port forwarding.
+
+[Install OpenSSH](https://docs.microsoft.com/en-us/windows-server/administration/openssh/openssh_install_firstuse#install-openssh-using-windows-settings) on the RobotStudio computer.
```
Comment (AndyZe) [inline docs/README.md]:
```suggestion
[Install "OpenSSH Server"](https://docs.microsoft.com/en-us/windows-server/administration/openssh/openssh_install_firstuse#install-openssh-using-windows-settings) on the RobotStudio computer.
```

File: docs/README.md [hunk 29]
```diff
@@ -87,18 +92,109 @@ The simulation will then try to connect with the ROS2 driver every few seconds.
...
+```ssh -L <DESTINATION_PORT>:localhost:<RWS_PORT> <ROBOTSTUDIO_USER>@<ROBOTSTUDIO_IP>```
+
+- Change `<DESTINATION_PORT>` to a port of your choice. Note that `sudo` is required to use ports 0-1023, as they are reserved. You may also need to change the `rws_port` parameter in the [robot description file](../robot_specific_config/abb_irb1200_support/urdf/irb1200.ros2_control.xacro) to match the port you selected. 
+- Change `<RWS_PORT>` to the port used for RWS on the RobotStudio machine, typically 80.
+- Change `<ROBOTSTUDIO_USER>` to the Windows username on the RobotStudio computer. This account must have a password.
+- Change `<ROBOTSTUDIO_IP>` to the IP of the RobotStudio computer.
+
+Enter the Windows login password for the `<ROBOTSTUDIO_USER>` account when prompted.
+
+This will allow you to access RWS on the ROS2 computer on `127.0.0.1:<DESTINATION_PORT>`. You will also need to change the `rws_ip` parameter in the [robot description file](../robot_specific_config/abb_irb1200_support/urdf/irb1200.ros2_control.xacro) to `127.0.0.1`.
```
Comment (AndyZe) [inline docs/README.md]:
```suggestion
This will allow you to access RWS on the ROS2 computer on `127.0.0.1:<DESTINATION_PORT>`. When launching ROS, use this value for robotstudio_ip: `127.0.0.1`.
```

**Review by AndyZe — APPROVED** [review]
Approved assuming you make the small changes above ^
